osv/impact_test.py

```diff
@@ -0,0 +1,149 @@
+
+  def setUp(self):
+    tests.reset_emulator()
+
+  @classmethod
+  def setUpClass(cls):
+    os.system('pkill -f datastore')
+    cls._ds_emulator = tests.start_datastore_emulator()
+
+    ndb_client = ndb.Client()
+    cls._ndb_context = ndb_client.context()
+    context = cls._ndb_context.__enter__()  # pylint: disable=unnecessary-dunder-call
+    context.set_memcache_policy(False)
+    context.set_cache_policy(False)
+
+  @classmethod
+  def tearDownClass(cls):
+    # TODO(ochang): Cleaner way of properly cleaning up processes.
+    os.system('pkill -f datastore')
+    cls._ndb_context.__exit__(None, None, None)  # pylint: disable=unnecessary-dunder-call
+
+  def test_update_single_page(self):
+    """Test update_affected_commits with a single page."""
+    commits = {
+        '00',
+        '01',
+        '02',
+    }
+
+    impact.update_affected_commits('BUG-1', commits, True, write_legacy=False)
+    affected_commits = list(models.AffectedCommits.query())
+    self.assertEqual(1, len(affected_commits))
+    affected_commits = affected_commits[0]
+
+    self.assertEqual('BUG-1-0', affected_commits.key.id())
+    self.assertEqual('BUG-1', affected_commits.bug_id)
+    self.assertEqual(0, affected_commits.page)
+    self.assertTrue(affected_commits.public)
+
+    self.assertCountEqual([
+        b'00',
+        b'01',
+        b'02',
+    ], [codecs.encode(commit, 'hex') for commit in affected_commits.commits])
+
+  def test_update_single_page_nonpublic(self):
+    """Test update_affected_commits with a single page."""
+    commits = {
+        '00',
+        '01',
+        '02',
+    }
+
+    impact.update_affected_commits('BUG-1', commits, False, write_legacy=False)
+    affected_commits = list(models.AffectedCommits.query())
+    self.assertEqual(1, len(affected_commits))
+    affected_commits = affected_commits[0]
+
+    self.assertEqual('BUG-1-0', affected_commits.key.id())
+    self.assertEqual('BUG-1', affected_commits.bug_id)
+    self.assertEqual(0, affected_commits.page)
+    self.assertFalse(affected_commits.public)
+
+    self.assertCountEqual([
+        b'00',
+        b'01',
+        b'02',
+    ], [codecs.encode(commit, 'hex') for commit in affected_commits.commits])
+
+  def test_update_multiple_pages(self):
+    """Test update_affected_commits with multiple page."""
+    # These pre-populated pages should be deleted.
+    for i in range(10):
+      models.AffectedCommits(
+          id=f'BUG-1-{i}', bug_id='BUG-1', commits=[], public=True,
+          page=i).put()
+
+    # These should not be deleted as they're for a different bug.
+    models.AffectedCommits(
+        id='BUG-2-0', bug_id='BUG-2', commits=[], public=True, page=0).put()
+
+    commits = {'%08d' % number for number in range(26000)}
+
+    impact.update_affected_commits('BUG-1', commits, True, write_legacy=False)
+    affected_commits = list(models.AffectedCommits.query())
+    self.assertEqual(4, len(affected_commits))
```

Comment:
I'm not able to follow why 4 is the expected result here.

I'm seeing 10 pre-populated commits going in for subsequent deletion (BUG-1-{0..9}). 
I'm seeing a single commit for a different bug going in (BUG-2-0).
I'm seeing 26,000 commits for a single bug (BUG-1)

Reply:
These are batched entities. 4 is the expected number that's batched.